File: Client/Frontend/WebExtensions/API/WebExtensionBrowserActionAPI.swift
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import Foundation
import Shared
import WebKit

class WebExtensionBrowserActionAPI: WebExtensionAPIEventDispatcher {
    override class var Name: String { return "browserAction" }

    enum Method: String {
        case setTitle
        case getTitle
        case setIcon
        case setPopup
        case getPopup
        case openPopup
        case setBadgeText
        case getBadgeText
        case setBadgeBackgroundColor
        case getBadgeBackgroundColor
        case setBadgeTextColor
        case getBadgeTextColor
        case enable
        case disable
        case isEnabled
    }
}

extension WebExtensionBrowserActionAPI: WebExtensionAPIConnectionHandler {
    func webExtension(_ webExtension: WebExtension, didReceiveConnection connection: WebExtensionAPIConnection) {
        guard let method = Method.init(rawValue: connection.method) else {
            connection.error("Unknown method: \(connection.method)")
            return
        }

        switch method {
        default:
            connection.error("Method not implemented: \(connection.method)")
        }
    }
}

File: Client/Frontend/WebExtensions/UserScripts/API/WebExtensionBrowserActionAPI.js
/* vim: set ts=2 sts=2 sw=2 et tw=80: */
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

const browserAction = {
  setTitle: noimpl("setTitle"),
  getTitle: noimpl("getTitle"),
  setIcon: noimpl("setIcon"),
  setPopup: noimpl("setPopup"),
  getPopup: noimpl("getPopup"),
  openPopup: noimpl("openPopup"),
  setBadgeText: noimpl("setBadgeText"),
  getBadgeText: noimpl("getBadgeText"),
  setBadgeBackgroundColor: noimpl("setBadgeBackgroundColor"),
  getBadgeBackgroundColor: noimpl("getBadgeBackgroundColor"),
  setBadgeTextColor: noimpl("setBadgeTextColor"),
  getBadgeTextColor: noimpl("getBadgeTextColor"),
  enable: noimpl("enable"),
  disable: noimpl("disable"),
  isEnabled: noimpl("isEnabled"),

  onClicked: new NativeEvent(SECURITY_TOKEN, WEB_EXTENSION_ID, "browser.browserAction.onClicked")
};

window.browser.browserAction = browserAction;
